src/index/hnsw/faiss_hnsw.cc

```diff
@@ -419,8 +444,298 @@ get_index_data_format(const faiss::Index* index) {
     return std::nullopt;
 }
 
+// cloned from IndexHNSW.cpp
+faiss::DistanceComputer*
+storage_distance_computer(const faiss::Index* storage) {
+    if (faiss::is_similarity_metric(storage->metric_type)) {
+        return new faiss::NegativeDistanceComputer(storage->get_distance_computer());
+    } else {
+        return storage->get_distance_computer();
+    }
+}
+
 }  // namespace
 
+// Contains an iterator state
+struct FaissHnswIteratorWorkspace {
+    // hnsw.
+    // this pointer is not owned.
+    const faiss::HNSW* hnsw = nullptr;
+
+    // nodes that we've already visited
+    faiss::cppcontrib::knowhere::Bitset visited_nodes;
+
+    // Computes distances.
+    //   This needs to be wrapped with a sign change.
+    std::unique_ptr<faiss::DistanceComputer> qdis;
+    // Computes refine distances (if refine is available).
+    //   This DOES NOT need to be wrapped with a sign change
+    std::unique_ptr<faiss::DistanceComputer> qdis_refine;
+
+    // for filtering out nodes
+    BitsetView bitset;
+
+    // accumulated alpha
+    float accumulated_alpha = 0;
+
+    // visitor
+    DummyVisitor graph_visitor;
+
```

Comment:
Can I ask what the `graph_visitor` is for? Do you have anything in mind for the future?

Reply:
`FederVisitor`